Cookbook/Proposals/SmarterDiffs.md

```diff
@@ -0,0 +1,97 @@
+        shouldShowGPAtHandStatusCard: true,
+        status: .loaded
+    )
+]))
+```
+
+This will result in diff
+
+```
+Found difference for child some:
+child :
+	shouldShowNHSBanner received: "true" expected: "false"
+```
+
+instead of
+
+```
+expected to equal <[State(firstName: "Brienne", lastName: "Skrilla", shouldShowNHSBanner: false, shouldShowHealthcheckBanner: false, shouldShowTestAndKitsBanner: false, shouldShowGPAtHandStatusCard: false, appointments: [], prescriptions: [], partnerLogoURL: nil, partnerLogoCaption: "", status: Babylon.HomeViewModelV2.Status.idle, isInsuranceValid: true), State(firstName: "Brienne", lastName: "Skrilla", shouldShowNHSBanner: false, shouldShowHealthcheckBanner: false, shouldShowTestAndKitsBanner: false, shouldShowGPAtHandStatusCard: false, appointments: [], prescriptions: [], partnerLogoURL: nil, partnerLogoCaption: "", status: Babylon.HomeViewModelV2.Status.loading, isInsuranceValid: true), State(firstName: "Brienne", lastName: "Skrilla", shouldShowNHSBanner: false, shouldShowHealthcheckBanner: true, shouldShowTestAndKitsBanner: true, shouldShowGPAtHandStatusCard: false, appointments: [], prescriptions: [], partnerLogoURL: nil, partnerLogoCaption: "", status: Babylon.HomeViewModelV2.Status.loaded, isInsuranceValid: true)]>, got <[State(firstName: "Brienne", lastName: "Skrilla", shouldShowNHSBanner: false, shouldShowHealthcheckBanner: false, shouldShowTestAndKitsBanner: false, shouldShowGPAtHandStatusCard: false, appointments: [], prescriptions: [], partnerLogoURL: nil, partnerLogoCaption: "", status: Babylon.HomeViewModelV2.Status.idle, isInsuranceValid: true), State(firstName: "Brienne", lastName: "Skrilla", shouldShowNHSBanner: false, shouldShowHealthcheckBanner: false, shouldShowTestAndKitsBanner: false, shouldShowGPAtHandStatusCard: false, appointments: [], prescriptions: [], partnerLogoURL: nil, partnerLogoCaption: "", status: Babylon.HomeViewModelV2.Status.loading, isInsuranceValid: true), State(firstName: "Brienne", lastName: "Skrilla", shouldShowNHSBanner: true, shouldShowHealthcheckBanner: true, shouldShowTestAndKitsBanner: true, shouldShowGPAtHandStatusCard: false, appointments: [], prescriptions: [], partnerLogoURL: nil, partnerLogoCaption: "", status: Babylon.HomeViewModelV2.Status.loaded, isInsuranceValid: true)]>
+```
+
+As can be seen from the diff it does not show in what exact array item the failure happened. This is something that we will need to adjust along with [this issue](https://github.com/krzysztofzablocki/Difference/issues/1). 
+
+Proposed way to integrate this feature is to copy the implementation (with proper author attribution) and adjust it to our needs, optionally we can make pull requests to the original library.
+
+## Impact on existing codebase
+
+This approach can be rolled out gradually and used for new test (Xcode templates will use a new assert) or tests can be refactored in batches by replacing existing assertions with a new assert.
```

Comment:
Given all issues with library, see my main review comment, I'd make it a non-default assertion. Something we can use when tests we're working on fails.
```swift
/// keep using `equal` for state tests
expect(states).to(equal([
    makeState(status: .idle),
    ...
)
// and if above fails we can switch to reflection based message for debugging or development
expect(states).to(equalWithDiff([
    makeState(status: .idle),
    ...
)
```

I'd keep it this way until we have more experience with it and have good idea in what areas we'd need to improve messages.

Reply:
Yes, probably.
The issues you highlighted seem all to relate to arrays to items, so we can try to mitigate them by applying diff algorithm after we apply reflection so that it will calculate mutations - if there are no mutations needed - assert passes, otherwise we see how things moved around.